File: JavaAlgorithm/src/main/java/algo/tzashinorpu/SpecialSubject/BinaryTree/maxProduct_1339.java
package algo.tzashinorpu.SpecialSubject.BinaryTree;

import algo.tzashinorpu.SpecialSubject.TreeNode;

public class maxProduct_1339 {
    int treeSum = 0;
    long res = 0;
    int mod = 1000000007;

    public int maxProduct(TreeNode root) {
        treeSum = getSum(root);
        getSum(root);
        return (int) (res % mod);
    }

    private int getSum(TreeNode root) {
        if (root == null) return 0;
        int l = getSum(root.left);
        int r = getSum(root.right);
        int rootSum = l + r + root.val;
        res = Math.max(res, ((long) rootSum * (treeSum - rootSum)));
        return rootSum;
    }
}

File: Clion/ZeroTrac/Tree/maxProduct_1339.cpp
//
// Created by TzashiNorpu on 10/27/2022.
//
#include "tree.h"

using namespace Tree;
#define ll long long

int Solution::maxProduct(TreeNode *root) {
  int sum = 0;
  ll mod = 1e9 + 7;
  function<void(TreeNode *)> hepler_1 = [&](TreeNode *node) -> void {
      if (!node) return;
      sum += node->val;
      hepler_1(node->left);
      hepler_1(node->right);
  };
  hepler_1(root);

  ll res = INT_MIN;
  function<ll(TreeNode *)> helper_2 = [&](TreeNode *node) -> ll {
      if (!node)return 0;
      ll l = helper_2(node->left);
      ll r = helper_2(node->right);
      ll v = l + r + node->val;
      res = max(res, (sum - v) * v);
      return v;
  };
  helper_2(root);
  return res % mod;
}